### PR TITLE
Replace Gulp with Fly

### PR DESCRIPTION
**Pull request by lukeed**
Fly and Gulp4 have _very_ similar goals; but Fly is already out of `beta`, faster, lighter, and more performant. Next.js [has already made the switch](https://github.com/zeit/next.js/pull/965)! 🎉 

I also added `fly-esnext`, which allows for the `async/await` flavor.

|                    | Gulp  | Fly   |
|--------------------|-------|-------|
| Total Dependencies | 662   | 590   |
| Time: `build` | 840ms   | 734ms  |
| Memory: `start`  | 43.8MB | 38.2MB |

**Slight Changes:**

- Sourcemaps are created as siblings to their source files
- Builds triggered by watcher are _slightly_ slower (17ms vs 26ms); can be easily fixed with a `fly-changed` plugin (TBD)

I'll totally understand if you choose not to migrate away from Gulp; or if you'd prefer to wait for the plugin before considering.

### CODE REVIEW

**Comment by dikaiosune** [comment]
Cool! I don't have any immediate objection to using fly, although I'm not familiar with the project. Could you link me to some information about what fly is?

Also, could you run `yarn && yarn run format` in the root repo directory?

**Comment by lukeed** [comment]
Hey, no worries! It's been around for over a year, but only recently came into its 2.0 form.

Here's the [README](https://github.com/flyjs/fly/blob/master/README.md), which explains the entire project & its mechanics pretty well in ~10 minutes. Let me know if you have questions.

I also ran the `format` command. Ran it again locally to check & there were no changes.

**Comment by dikaiosune** [comment]
Cool! It's on my list to take a look at when a little bit of the issue churn slows down. It looks like the fly scripts are using tabs, but prettier *should* have reformatted them to spaces.

**Comment by lukeed** [comment]
Ah, gotcha. It's because you have `{src,template}` included in your glob. I'll just transform to spaces so you don't have to evolve a slightly more complicated glob. Plus I'm not sure if you can even use multiple glob patterns & ignores with `prettier` yet. 🤔

File: react-native-scripts/package.json [hunk 1]
```diff
@@ -40,12 +40,12 @@
...
-    "gulp-shell": "^0.5.2",
-    "gulp-sourcemaps": "^1.6.0",
-    "rimraf": "^2.5.2"
+    "fly": "^2.0.5",
+    "fly-babel": "^2.1.1",
+    "fly-clear": "^1.0.1",
+    "fly-esnext": "^2.0.0",
+    "fly-shell": "^2.1.0",
+    "fly-watch": "^1.1.1",
+    "gulp-shell": "^0.5.2"
```
Comment (dikaiosune) [inline react-native-scripts/package.json]: Does fly use gulp's shell module?

File: react-native-scripts/flyfile.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+export async function clean(fly) {
+  await fly.clear(paths.build);
+}
+
+export async function build(fly) {
+  await fly.start('babel');
+}
+
+export async function publish(fly) {
+  await fly.serial(['clear', 'babel']).shell('npm publish');
```
Comment (dikaiosune) [inline react-native-scripts/flyfile.js]: In fly will a serial task abort midway if a subtask fails?

File: react-native-scripts/package.json (hunk 1 above)
Comment (lukeed) [inline react-native-scripts/package.json]: Oh, no. Must have slipped in during one of my copy-pastes.

File: react-native-scripts/flyfile.js (hunk 2 above)
Comment (lukeed) [inline react-native-scripts/flyfile.js]: Yes

**Comment by dikaiosune** [comment]
I'm trying this out, and it looks like fly needs to be called from package.json scripts to work with the async/await flavor:

```
$ fly babel
/Users/adam/code/flycrna/create-react-native-app/flyfile.js:9
export default async function (fly) {
^^^^^^
SyntaxError: Unexpected token export
    at new Fly (/Users/adam/.nvm/versions/node/v7.7.1/lib/node_modules/fly/lib/fly.js:21:39)
    at Object.<anonymous> (/Users/adam/.nvm/versions/node/v7.7.1/lib/node_modules/fly/lib/cli/spawn.js:35:9)
    at Generator.next (<anonymous>)
    at Generator.tryCatcher (/Users/adam/.nvm/versions/node/v7.7.1/lib/node_modules/fly/node_modules/bluebird/js/release/util.js:16:23)
    at PromiseSpawn._promiseFulfilled (/Users/adam/.nvm/versions/node/v7.7.1/lib/node_modules/fly/node_modules/bluebird/js/release/generators.js:97:49)
    at Promise._settlePromise (/Users/adam/.nvm/versions/node/v7.7.1/lib/node_modules/fly/node_modules/bluebird/js/release/promise.js:574:26)
    at Promise._settlePromise0 (/Users/adam/.nvm/versions/node/v7.7.1/lib/node_modules/fly/node_modules/bluebird/js/release/promise.js:614:10)
    at Promise._settlePromises (/Users/adam/.nvm/versions/node/v7.7.1/lib/node_modules/fly/node_modules/bluebird/js/release/promise.js:693:18)
    at Async._drainQueue (/Users/adam/.nvm/versions/node/v7.7.1/lib/node_modules/fly/node_modules/bluebird/js/release/async.js:133:16)
    at Async._drainQueues (/Users/adam/.nvm/versions/node/v7.7.1/lib/node_modules/fly/node_modules/bluebird/js/release/async.js:143:10)
    at Immediate.Async.drainQueues (/Users/adam/.nvm/versions/node/v7.7.1/lib/node_modules/fly/node_modules/bluebird/js/release/async.js:17:14)
```

Can you add the scripts to package.json so that all commands are available via `npm run`/`yarn`?

**Comment by lukeed** [comment]
Okay. Right now you can do that by running `npm start babel` simply because it's passing all task names thru that `fly` bin reference.

Of course, that's not ideal. But I matched the existing Gulp configuration, which only had `start` paired to `gulp`.

**Comment by dikaiosune** [comment]
Ah fair enough. Cool, I can tweak as needed then.

**Comment by lukeed** [comment]
Cool! Let me know if you have any questions while tweaking. More than happy to help!

Also, _just_ missed the mark. Just pushed a commit for removing `gulp-shell` 😅

**Comment by dikaiosune** [comment]
All good! Figured I was going to tinker a bit anyways 😃

**Pull request merged**